src/apple/macos/ffi.rs

```diff
@@ -154,3 +177,133 @@ mod io_service {
+pub(crate) const HID_DEVICE_PROPERTY_PRIMARY_USAGE_PAGE: &str = "PrimaryUsagePage";
+pub(crate) const HID_DEVICE_PROPERTY_MAX_INPUT_REPORT_SIZE: &str = "MaxInputReportSize";
+pub(crate) const HID_DEVICE_PROPERTY_MAX_OUTPUT_REPORT_SIZE: &str = "MaxOutputReportSize";
+pub(crate) const HID_DEVICE_PROPERTY_REPORT_ID: &str = "ReportID";
+
+pub(crate) const kHIDPage_AppleVendor: i32 = 0xff00;
+pub(crate) const kHIDUsage_AppleVendor_TemperatureSensor: i32 = 0x0005;
+
+pub(crate) fn matching(page: i32, usage: i32) -> CFDictionaryRef {
+    unsafe {
+        let primary_usage_page = CString::new(HID_DEVICE_PROPERTY_PRIMARY_USAGE_PAGE).unwrap();
+        let primary_usage = CString::new(HID_DEVICE_PROPERTY_PRIMARY_USAGE).unwrap();
+
+        let keys = [
+            CFStringCreateWithCString(
+                null() as *const _, 
+                primary_usage_page.as_ptr(), 
+                0
+            ),
+            CFStringCreateWithCString(
+                null() as *const _, 
+                primary_usage.as_ptr(), 
+                0
+            ),
+        ];
+
+        let nums = [
+            CFNumberCreate(
+                null(), 
+                kCFNumberSInt32Type, 
+                &page as *const _ as *const libc::c_void
+            ),
+            CFNumberCreate(
+                null(), 
+                kCFNumberSInt32Type, 
+                &usage as *const _ as *const libc::c_void
+            ),
+        ];
+
+        CFDictionaryCreate(
+            null(), 
+            &keys as *const _ as *const _,
+            &nums as *const _ as *const _, 
+            2,
+            &kCFTypeDictionaryKeyCallBacks,
+            &kCFTypeDictionaryValueCallBacks
+        )
+    }
+}
+
+pub fn get_product_names(sensors: CFDictionaryRef) -> Vec<String> {
+    unsafe {
+        let system = IOHIDEventSystemClientCreate(kCFAllocatorDefault);
+
+        let _rv = IOHIDEventSystemClientSetMatching(system, sensors);
+
+        let services = IOHIDEventSystemClientCopyServices(system);
+
+        let count = CFArrayGetCount(services);
+
+        let key_ref = CFStringCreateWithBytes(
+            kCFAllocatorDefault, 
+            HID_DEVICE_PROPERTY_PRODUCT.as_ptr(), 
+            HID_DEVICE_PROPERTY_PRODUCT.len() as CFIndex, 
+            kCFStringEncodingUTF8,
+            false as Boolean
+        );
+
+        let mut names = Vec::with_capacity(count as usize);
+
+        for i in 0..count {
+            let service = CFArrayGetValueAtIndex(services, i);
+            let name = IOHIDServiceClientCopyProperty(service as *const _, key_ref);
+            if name != null() {
+                // CFArrayAppendValue(array, name as *const _);
+                let name_ptr = CFStringGetCStringPtr(name as *const _, kCFStringEncodingUTF8);
+                let name = CStr::from_ptr(name_ptr).to_string_lossy();
+                names.push(name);
+            } 
+        }
+
+        names.into_iter().map(|i| i.to_string()).collect::<Vec<_>>()
```

Comment:
Instead of recreating a new `Vec` here, please replace `names.push(name)` with `name.push(name.to_string())`.